### PR TITLE
Modernise macOS cross-compilation toolchain

### PR DESCRIPTION
**Pull request by str4d**
Cherry-picked from the following upstream PRs:

- bitcoin/bitcoin#13617
  - Excluding the QT GUI changes.
- bitcoin/bitcoin#17550
- bitcoin/bitcoin#16392
  - Excluding the QT GUI changes.
- bitcoin/bitcoin#18589
- bitcoin/bitcoin#19240
- bitcoin/bitcoin#19407
- bitcoin/bitcoin#17919
  - Only the ancillary changes, not the `FORCE_USE_SYSTEM_CLANG` change.
- bitcoin/bitcoin#19530

After these changes, macOS versions earlier than 10.12 are no longer supported.

To cross-compile for macOS:
- Follow the instructions in `contrib/macdeploy/README.md` to generate
  `Xcode-11.3.1-11C505-extracted-SDK-with-libcxx-headers.tar.gz`
  (requires an Apple Developer Account)
- Extract it into `depends/SDKs` (creating that folder first if it does not exist)
- `HOST=x86_64-apple-darwin16 ./zcutil/build.sh`

### CODE REVIEW

**Comment by daira** [comment]
@zkbot try

**Comment by str4d** [comment]
The `ubuntu1804` builder had a transient failure, and otherwise the only failing builder is the `debian9_darwin11` cross-compiler as expected (because it is using the old macOS SDK).

**Comment by rex4539** [comment]
This compiles without errors on native macOS Catalina.

As a side effect, it also fixes https://github.com/zcash/zcash/issues/4309.

File: doc/release-notes.md [hunk 1]
```diff
@@ -4,3 +4,6 @@ release-notes at release time)
 Notable changes
 ===============
 
+Compatibility
+-------------
+macOS versions earlier than 10.12 are no longer supported.
```
Comment (daira) [inline doc/release-notes.md:9]:
```suggestion
macOS versions earlier than 10.12 (Sierra) are no longer supported.
```

**Comment by str4d** [comment]
@zkbot r+

**Comment by daira** [comment]
Post-hoc ACK for the comment changes since my last review.

**Pull request merged**